### PR TITLE
Add typescript defintions

### PR DESCRIPTION
**Pull request by bilby91**

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,2633 @@
+{
```
Comment (dougwilson) [inline package-lock.json]: Please remove this file 👍

File: typings/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import * as express from "express"
```
Comment (dougwilson) [inline typings/index.d.ts]: So this module works just fine without Express. For example, using the Node.js HTTP server, connect, and more (see the examples in the README). Can the typings not be specific to Express.js, please?

File: package.json [hunk 3]
```diff
@@ -9,26 +9,32 @@
...
-    "supertest": "1.1.0"
+    "supertest": "1.1.0",
+    "typescript": "2.8.1"
   },
   "files": [
     "LICENSE",
     "HISTORY.md",
-    "index.js"
+    "index.js",
+    "typings/index.d.ts"
```
Comment (dougwilson) [inline package.json]: In other projects I've gotten typings, they are just in the root as `index.d.ts`. Any reason they need to be in their own folder, or can they just be in the root? A major benefit of it being in the root means that you don't need to have the `typings` key at all (just like `main` is not needed when the entry point is `index.js`).

Comment (bilby91) [reply package.json]: Sure, where would you put the test file ? `test/test.ts` sound good ?

File: package-lock.json (hunk 1 above)
Comment (bilby91) [reply package-lock.json]: Sure.

File: package.json (hunk 3 above)
Comment (dougwilson) [reply package.json]: Test file in `test` folder makes sense to me 👍

File: typings/index.d.ts (hunk 2 above)
Comment (bilby91) [reply typings/index.d.ts]: Made the more generic now

File: test/test.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import * as connect from "connect"
+
+const app = connect()
+
+// Using application
+vhost("*.example.com", app)
+
+// Using request handler
+vhost("*.example.com",  (req, res, next) => {
+  if (req.vhost instanceof Array) {
```
Comment (dougwilson) [inline test/test.ts]: Is this ever true? I don't believe req.vhost is ever an array, right?

File: package.json [hunk 5]
```diff
@@ -18,19 +20,23 @@
...
+    "typescript": "2.8.1"
   },
   "files": [
     "LICENSE",
     "HISTORY.md",
-    "index.js"
+    "index.js",
+    "index.d.ts"
   ],
+  "typings": "index.d.ts",
```
Comment (dougwilson) [inline package.json]: You can remove this as mentioned above, since it is now the standard name :+1:

**Comment by dougwilson** [comment]
Please tell me when you're done force pushing. Every time I try to make a comment the page reload and everything I typed is lost.

**Comment by bilby91** [comment]
@dougwilson Sorry for that, it was tricky to get it working for old version of npm that don't support scoped modules. I think travis should fine now.

File: package.json (hunk 5 above)
Comment (bilby91) [reply package.json]: Cool, though you needed to make explicit either way.

Comment (dougwilson) [reply package.json]: No you don't. Please remove it.

File: test/test.ts (hunk 4 above)
Comment (bilby91) [reply test/test.ts]: I based this on the README (https://github.com/expressjs/vhost/blame/master/README.md#L48)

Comment (dougwilson) [reply test/test.ts]: There is no instanceof check in the readme.

File: package.json (hunk 5 above)
Comment (bilby91) [reply package.json]: Sure, will remove it!

File: test/test.ts (hunk 4 above)
Comment (bilby91) [reply test/test.ts]: Got your point, my bad that I was only thinking in typings. I will change to the following and the problem should be solved.

```
  interface IVHost {
    [key: number]: string
    host: string,
    hostname: string,
    length: number
  }
```

Thanks!

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import * as http from "http"
+
+declare namespace vhost {
+  interface RequestHandler {
+    (req: http.IncomingMessage, res: http.ServerResponse, next: Function): void
+  }
+}
+
+declare function vhost(host: string, handler: vhost.RequestHandler): vhost.RequestHandler
```
Comment (dougwilson) [inline index.d.ts]: Ok, this was my last comment from yesterday: the `host` parameter also accepts a regular expression here. And should these parameters be named the same as they are in the docs and underlying JS? I assume users will see these names in TS errors?

**Comment by bilby91** [comment]
I think it should be good now @dougwilson

**Comment by bilby91** [comment]
Ping @dougwilson

**Comment by dougwilson** [comment]
I'm trying to get someone on my side who knows TypeScript to review this still.

**Comment by Watchieee** [comment]
I copied the .d.ts file into the installed module within my TS project and it seems to work fine. Would love to see this pr getting merged soon.

**Comment by imjakechapman** [comment]
This has been a year, can this get merged? I'll buy someone a beer, or fly and hug you. K thanks :)

**Comment by kistora** [comment]
Ping @dougwilson

**Comment by PabloSzx** [comment]
i made a fork refactored in typescript https://github.com/PabloSzx/vhost-ts https://www.npmjs.com/package/vhost-ts but for now only handles express types *(open for PRs)*

it also includes #21

**Comment by bilby91** [comment]
Ping @dougwilson

**Comment by bilby91** [comment]
I'm doing some cleanup of my PRs and I'm no longer going to work on this one.

**Pull request closed without merging**